solidity/contracts/midas/WithdrawalQueue.sol

```diff
@@ -0,0 +1,289 @@
+    /// @param _midasAllocator Address of the Midas Allocator contract.
+    /// @param _tbtcVault Address of the tBTC Vault contract.
+    /// @param _acreBTC Address of the acreBTC contract.
+    function initialize(
+        address _tbtc,
+        address _vault,
+        address _midasAllocator,
+        address _tbtcVault,
+        address _acreBTC
+    ) public initializer {
+        __MaintainableOwnable_init(msg.sender);
+
+        if (_tbtc == address(0)) {
+            revert ZeroAddress();
+        }
+        if (address(_vault) == address(0)) {
+            revert ZeroAddress();
+        }
+        if (_tbtcVault == address(0)) {
+            revert ZeroAddress();
+        }
+        if (_acreBTC == address(0)) {
+            revert ZeroAddress();
+        }
+
+        tbtc = ITBTCToken(_tbtc);
+        vault = IVault(_vault);
+        tbtcVault = _tbtcVault;
+        acrebtc = acreBTC(_acreBTC);
+
+        vaultSharesToken = IERC20(vault.share());
+        if (address(vaultSharesToken) == address(0)) {
+            revert ZeroAddress();
+        }
+
+        if (_midasAllocator == address(0)) {
+            revert ZeroAddress();
+        }
+
+        midasAllocator = MidasAllocator(_midasAllocator);
+    }
+
+    function requestRedeem(
+        uint256 _shares,
+        address _receiver,
+        uint256 _exitFeeInAssets
+    ) external onlyAcreBTC {
+        uint256 tbtcAmount = acrebtc.convertToAssets(_shares);
+        uint256 midasShares = vault.convertToShares(tbtcAmount);
+        midasAllocator.withdraw(midasShares);
+        vaultSharesToken.approve(address(vault), midasShares);
+        acrebtc.burn(_shares);
+        if (_exitFeeInAssets > 0) {
+            uint256 exitFeeShares = vault.convertToShares(_exitFeeInAssets);
+            vault.requestRedeem(exitFeeShares, acrebtc.treasury());
+            midasShares -= exitFeeShares;
+        }
+        vault.requestRedeem(midasShares, _receiver);
+    }
+
+    /// @notice Requests a redemption with extra bridge data
+    /// @param _shares Amount of shares to withdraw.
+    /// @param _walletPubKeyHash Wallet public key hash.
+    /// @param _redeemer The original redeemer address.
+    /// @return requestId The ID of the withdrawal request.
+    function requestRedeemAndBridge(
+        uint256 _shares,
+        bytes20 _walletPubKeyHash,
+        address _redeemer
+    ) external onlyAcreBTC returns (uint256 requestId) {
+        uint256 tbtcAmount = acrebtc.convertToAssets(_shares);
+        uint256 midasShares = vault.convertToShares(tbtcAmount);
+        midasAllocator.withdraw(midasShares);
+        acrebtc.burn(_shares);
+        vaultSharesToken.approve(address(vault), midasShares);
+        uint256 midasRequestId = vault.requestRedeem(
+            midasShares,
+            address(this)
+        );
+
+        requestId = count;
+        withdrawalRequests[requestId] = WithdrawalRequest({
+            redeemer: _redeemer,
+            shares: midasShares,
```

Comment:
Let's be specific that these are the Midas Vault shares, so it's not confused with the acreBTC shares.
```suggestion
            midasShares: midasShares,
```

Reply:
But do we need to store it at all? I don't see it being used for finalization. Maybe emitting in event is enough?